src/octree/batch_iterator.rs

```diff
@@ -144,29 +154,46 @@ impl<'a> BatchIterator<'a> {
     }
 
     /// compute a function while iterating on a batch of points
-    pub fn try_for_each_batch<F>(&mut self, mut func: F) -> Result<()>
+    pub fn try_for_each_batch<F>(&mut self, func: F) -> Result<()>
     where
         F: FnMut(PointData) -> Result<()>,
     {
-        //TODO(catevita): mutable function parallelization
+        let octree = self.octree;
+        // nodes iterator: retrieve nodes
+        let node_id_iterator = octree.nodes_in_location(self.point_location);
+
         let local_from_global = self
             .point_location
             .global_from_local
             .clone()
             .map(|t| t.inverse());
-        let mut point_stream = PointStream::new(self.batch_size, local_from_global, &mut func);
-        // nodes iterator: retrieve nodes
-        let node_id_iterator = self.octree.nodes_in_location(self.point_location);
-        // operate on nodes
-        for node_id in node_id_iterator {
-            let point_iterator = self.octree.points_in_node(self.point_location, node_id);
-            for point in point_iterator {
-                point_stream.push_point_and_callback(point)?;
+        // operate on nodes: one thread for each node
+        crossbeam::scope(|s| {
+            let (tx, rx) = crossbeam::channel::bounded::<PointData>(2);
+            for node_id in node_id_iterator {
+                let tx = tx.clone();
+                let local_from_global = local_from_global.clone();
```

Comment:
Nit: We could just use a reference to lo local_from_global and self.point_location and avoid the clone.

Reply:
ok!

Reply:
a reference does not work for the isometry though..